src/solver/src/simulation.C

```diff
@@ -487,12 +497,57 @@ namespace GRINS
          it++ )
       {
         std::tr1::shared_ptr<Physics> physics = system->get_physics( it->first );
-      
+
         physics->attach_dirichlet_bound_func( it->second );
       }
     return;
   }
 
+  void Simulation::init_adjoint_solve( const GetPot& input, bool output_adjoint )
+  {
+    // Check if we're doing an adjoint solve
+    _do_adjoint_solve = this->check_for_adjoint_solve(input);
+
+    const libMesh::DifferentiableQoI* raw_qoi = _multiphysics_system->get_qoi();
+    const CompositeQoI* qoi = dynamic_cast<const CompositeQoI*>( raw_qoi );
+
+    // If we are trying to do an adjoint solve without a QoI, that's an error
+    if( _do_adjoint_solve && qoi->n_qois() == 0 )
+      {
```

Comment:
> I think this will fail. Right now if we don't have any QoIs we don't even bother building a CompositeQoI and then the dynamic_cast will return NULL.

Spot on, just tried it. Good catch.